### ISSUE TITLE
Document preservation / patching tables in-line?

### ISSUE THREAD

**Issue by DivineDominion** [MarkCodable/MarkCodable#36]
Currently, there's no support for any content besides tables.

For the value extraction, this works perfectly fine: the table in the following test case is extracted properly:

```swift
func testDocument() throws {
    let markdown = """
    # Heading
    What is this?

    |value|
    |-----|
    |2    |
    |4    |

    A table!
    """

    struct ValueHolder: Codable, Equatable {
        let value: Int
    }

    let encoder = MarkEncoder()
    let decoder = MarkDecoder()

    let tableContents = [ValueHolder(value: 2), ValueHolder(value: 4)]
    XCTAssertEqual(try encoder.encode(tableContents), markdown) // ❎
    XCTAssertEqual(try decoder.decode([ValueHolder].self, string: markdown), tableContents) // ✅
}
```

To get data from withing documents, this is excellent.

But what about editing documents?

My proposal is to store the rest of the Markdown string as well and put the document back together in the end. That's a different kind of serialization.

Is this beyond the scope of MarkCodable?

**Comment by DivineDominion** [comment]
Somewhat related: Supporting multiple tables in a single file.

The GFM specs which `swift-markdown` is based on sadly don't support table names. [MultiMarkdown does this to reference tables by ID](https://fletcher.github.io/MultiMarkdown-6/syntax/#tables), which would've been very useful. (Dang.)

**Comment by icanzilb** [comment]
> My proposal is to store the rest of the Markdown string as well and put the document back together in the end. That's a different kind of serialization.
> 
> Is this beyond the scope of MarkCodable?

Dang, I can see how that could be useful... the problem, as I see it, is that for the codec there is no notion of "document". The encoder and decoder are separate entities because they don't track the history of the values they work with. I.e. to keep them properly separated I would think there would be a `Document` object or a `MarkdownMultipartDocument` that would preserve the document information and use the codec for persistence.

In other words — I'm a little afraid to expand this from purely encoding/decoding values into a type that implements logic, loads semantic meaning into markdown (kind of a DocC of its own). Not that I don't think it should happen but I'd rather see the basics work correctly for a fact before expanding. I hope that sounds reasonable

**Comment by DivineDominion** [comment]
I was initially imagining shipping a `Document` type or similar as a container to parse whole documents, but I now realize that you're right: maybe it's not a good idea to put that functionality into this package at all. This package can focus on tabular data encoding/decoding, and another package can _use this_ to make whole documents codable. 👍
